csharp/src/Apache.Arrow.Adbc/Interop/LoadDriver.cs

```diff
@@ -0,0 +1,508 @@
+                NativeAdbcError error = new NativeAdbcError();
+                byte result = init(1000000, ref driver, ref error);
+                return new NativeDriver { /* driverHandle = library, */ driver = driver };
+            }
+        }
+
+        /// <summary>
+        /// Loads an <see cref="AdbcDriver"/> from the file system.
+        /// </summary>
+        /// <param name="file">The path to the file</param>
+        /// <returns></returns>
+        public static AdbcDriver Load(string file)
+        {
+            if (file[0] == '/')
+            {
+                return new AdbcDriverNative(MacInterop.GetDriver(file).driver);
+            }
+            else
+            {
+                return new AdbcDriverNative(WindowsInterop.GetDriver(file).driver);
+            }
+        }
+
+        /// <summary>
+        /// Native implementation of <see cref="AdbcDriver"/>
+        /// </summary>
+        sealed class AdbcDriverNative : AdbcDriver
+        {
+            private NativeAdbcDriver _nativeDriver;
+
+            public AdbcDriverNative(NativeAdbcDriver nativeDriver)
+            {
+                _nativeDriver = nativeDriver;
+            }
+
+            /// <summary>
+            /// Opens a database
+            /// </summary>
+            /// <param name="parameters"></param>
+            /// <returns></returns>
+            public unsafe override AdbcDatabase Open(Dictionary<string, string> parameters)
+            {
+                NativeAdbcDatabase nativeDatabase = new NativeAdbcDatabase();
+                using (ErrorHelper error = new ErrorHelper())
+                {
+                    error.Call(
+                        Marshal.GetDelegateForFunctionPointer<DatabaseFn>((IntPtr)_nativeDriver.DatabaseNew),
+                        ref nativeDatabase);
+
+                    DatabaseSetOption setOption = Marshal.GetDelegateForFunctionPointer<DatabaseSetOption>((IntPtr)_nativeDriver.DatabaseSetOption);
+                    if (parameters != null)
+                    {
+                        foreach (KeyValuePair<string, string> pair in parameters)
+                        {
+                            error.Call(setOption, ref nativeDatabase, pair.Key, pair.Value);
+                        }
+                    }
+                    error.Call(Marshal.GetDelegateForFunctionPointer<DatabaseFn>((IntPtr)_nativeDriver.DatabaseInit), ref nativeDatabase);
+                }
+
+                return new AdbcDatabaseNative(_nativeDriver, nativeDatabase);
+            }
+
+            public unsafe override void Dispose()
+            {
+                if (_nativeDriver.release != null)
+                {
+                    using (ErrorHelper error = new ErrorHelper())
+                    {
+                        try
+                        {
+                            error.Call(Marshal.GetDelegateForFunctionPointer<DriverRelease>((IntPtr)_nativeDriver.release), ref _nativeDriver);

```

Comment:
You don't need to / shouldn't use `Marshal.GetDelegateForFunctionPointer` here. Instead you can invoke the function like:

```C#
fixed (NativeAdbcDriver* driver = &nativeDriver)
{
    _nativeDriver.release(driver);
}
```

You'll probably have to rewrite your ErrorHelper for a new pattern. But there's no reason to allocate a managed object (a Delegate) just to call this function - or any native function.

Reply:
done in the latest push